### PR TITLE
[RLlib] EnvRunner API

### PR DESCRIPTION
**Pull request by sven1977**
**New EnvRunner API**:
- This PR introduces a new simple API for env runners and environment sample collection workers.
- The current env-runner solution (RolloutWorker + Sampler + EnvRunner(v2?)) is overloaded/convoluted and therefore hard to understand for users, let alone customize.
- The existing classes will NOT be affected by this API change (at least not beyond a few lines of code changed in each). However, `EnvRunner` is designed to serve all future Algorithm development needs from here on, e.g. DreamerV3, and will replace (in the mid run): `RolloutWorker`, `SyncSampler`, `AsyncSampler`, `InputReader`, `VectorEnv`, `BaseEnv`, `EnvRunnerV2`, `Episode` and `EpisodeV2`, `SampleCollector`.
- The main components of `EnvRunner` are: c'tor with config, Ray remote'able (fault tolerant within a `FaultTolerantActorManager`, a central `sample()` method.

An example EnvRunner implementation is part of this PR (including tests with async/parallel gym.vector.Env).
The upcoming DreamerV3 implementation in RLlib will use `EnvRunner` as well.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/env_runner.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import platform
+from typing import Any, Dict, Optional, Union
+
+import ray
+from ray.rllib.utils.actor_manager import FaultAwareApply
+from ray.rllib.utils.annotations import ExperimentalAPI
+
+
+@ExperimentalAPI
+class EnvRunner(FaultAwareApply, metaclass=abc.ABCMeta):
```
Comment (kouroshHakha) [inline rllib/env/env_runner.py:11]: When doing these refactors, can you clarify which components are copy / pasted from the past APIs? For example I don't know why we need get_node_ip() API etc. for this class. Is that really needed? or is that just copy / pasted from what we had before?

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
We already have an EnvRunner class (that holds a RolloutWorker).
Sadly, we already also have a SampleCollector class.
Do we want to use this same API for offline learning as well? If so, EnvRunner would not reflect that us case well.

File: rllib/env/testing/single_agent_gym_env_runner.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    def sample(
+        self,
+        *,
+        num_timesteps: Optional[int] = None,
+        num_episodes: Optional[int] = None,
+        explore: bool = True,
+        random_actions: bool = False,
+        force_reset: bool = False,
+        **kwargs,
+    ) -> Tuple[List[Episode], List[Episode]]:
```
Comment (kouroshHakha) [inline rllib/env/testing/single_agent_gym_env_runner.py:47]: Can you add a docstring on what the expected output type is? and what each element in the tuple represent?

File: rllib/env/testing/single_agent_gym_env_runner.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                    self._split_by_env(truncateds),
+                )
+            ):
+                # Episode is done (terminated or truncated).
+                if term or trunc:
+                    # Finish the episode object with the actual terminal observation
+                    # stored in the info dict (`o` is already the reset obs of the new
+                    # episode).
+                    self._episodes[i].add_timestep(
+                        infos["final_observation"][i], a, r, is_terminated=True
```
Comment (kouroshHakha) [inline rllib/env/testing/single_agent_gym_env_runner.py:111]: do all gymnasium envs follow this convention now? o returned by `.step()` is the reset of the next env? Can we comment that detail here?

Comment (kouroshHakha) [reply rllib/env/testing/single_agent_gym_env_runner.py:111]: So we should not necessarily hit `.reset()` anymore?

File: rllib/env/tests/test_single_agent_gym_env_runner.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    def test_sample(self):
+        config = (
+            AlgorithmConfig().environment("CartPole-v1")
+            # Vectorize x2 and by default, rollout 64 timesteps per individual env.
+            .rollouts(num_envs_per_worker=2, rollout_fragment_length=64)
+        )
+        env_runner = SingleAgentGymEnvRunner(config=config)
+
+        # Sample 10 episodes 10 times.
```
Comment (kouroshHakha) [inline rllib/env/tests/test_single_agent_gym_env_runner.py:25]: we currently don't support sampling based on `num_episodes` right?

File: rllib/evaluation/rollout_worker.py [hunk 5]
```diff
@@ -163,7 +162,7 @@ def _update_env_seed_if_necessary(
 
 
 @DeveloperAPI
-class RolloutWorker(ParallelIteratorWorker, FaultAwareApply):
+class RolloutWorker(ParallelIteratorWorker, EnvRunner):
```
Comment (kouroshHakha) [inline rllib/evaluation/rollout_worker.py:165]: oh ok. So RolloutWorker is a special EnvRunner. I see now why you made certain choices. But I think still the parent class may not need get_node_ip or those other APIs.

File: rllib/evaluation/rollout_worker.py [hunk 6]
```diff
@@ -857,43 +653,16 @@ def wrap(env):
...
         is_healthy = self.policy_map and self.input_reader and self.output_writer
         assert is_healthy, (
             f"RolloutWorker {self} (idx={self.worker_index}; "
             f"num_workers={self.num_workers}) not healthy!"
         )
 
-    @DeveloperAPI
+    @override(EnvRunner)
     def sample(self) -> SampleBatchType:
-        """Returns a batch of experience sampled from this worker.
```
Comment (kouroshHakha) [inline rllib/evaluation/rollout_worker.py:878]: I think we need to keep the docstring around, right?

File: rllib/env/env_runner.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+        Raises:
+            AssertionError: If the EnvRunner Actor has NOT been properly initialized.
+        """
+
+    @abc.abstractmethod
+    def sample(
+        self,
+        *,
+        num_timesteps: Optional[int] = None,
```
Comment (kouroshHakha) [inline rllib/env/env_runner.py:96]: Is this API needed? I noticed that RW completely override the signature.

File: rllib/evaluation/worker_set.py [hunk 8]
```diff
@@ -1003,6 +974,29 @@ def _valid_module(cls, class_path):
...
+
+    @Deprecated(
+        old="remote_workers()",
+        help=(
+            "Accessing the list of remote workers directly through "
+            "remote_workers() is strongly discouraged. "
+            "Please try to use one of the foreach accessors "
+            "that is fault tolerant. "
+        ),
+        error=False,
```
Comment (kouroshHakha) [inline rllib/evaluation/worker_set.py:999]: should we turn the error=True on?

**Review by kouroshHakha — COMMENTED** [review]
I really like the how light weight this is. We can basically branch off from newly introduce parent of Rollout Workers and keep the old stack functional for a foreseeable future. I have a couple of questions on the EnvRunner. I feel like the currenly imposed API on it is heavier than what it needs to be, specifically I don't get why we still need the following:

```
get_host()
get_node_ip()
```

Also the args of `sample()` seem too much degrees of freedom. Why not only use configs that are already validated?

File: rllib/env/env_runner.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        return {}
+
+    def set_state(self, state: Dict[str, Any]) -> None:
+        """Restores this EnvRunner's state from the given state dict.
+
+        Args:
+            state: The state dict to restore the state from.
+
+        Examples:
+            >>> from ray.rllib.evaluation.rollout_worker import RolloutWorker
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/env_runner.py:145]: Great catch, will fix.

File: rllib/evaluation/worker_set.py (hunk 8 above)
Comment (sven1977) [reply rllib/evaluation/worker_set.py:999]: Let's do it and see what happens :)

File: rllib/evaluation/rollout_worker.py (hunk 6 above)
Comment (sven1977) [reply rllib/evaluation/rollout_worker.py:878]: Maybe in this case, yes. I was thinking since I moved it into the base class that I wouldn't need this here. But the behavior might be different still.

File: rllib/evaluation/rollout_worker.py (hunk 5 above)
Comment (sven1977) [reply rllib/evaluation/rollout_worker.py:165]: Correct:
Base: EnvRunner --> RolloutWorker
                              \-> DreamerEnvRunner

True, maybe these extra APIs we don't need (leave them in RW). Lemme think about these some more.

File: rllib/env/testing/single_agent_gym_env_runner.py (hunk 3 above)
Comment (sven1977) [reply rllib/env/testing/single_agent_gym_env_runner.py:111]: Correct, with gymnasium.vector, you'll only have to reset once at the very beginning. After that, each sub-env resets itself automatically. Actually, there is a setting where you can switch off that behavior, but by default, it's on.
I'll add more explanations.

File: rllib/env/env_runner.py (hunk 1 above)
Comment (sven1977) [reply rllib/env/env_runner.py:11]: will do!

File: rllib/evaluation/worker_set.py [hunk 10]
```diff
@@ -1003,6 +974,29 @@ def _valid_module(cls, class_path):
...
-    @Deprecated(new="WorkerSet.is_policy_to_train([pid], [batch]?)", error=True)
-    def trainable_policies(self):
-        pass
+    @property
+    @Deprecated(
+        old="_remote_workers",
+        help=(
+            "Accessing remote workers directly through "
+            "_remote_workers is strongly discouraged. "
+            "Please try to use one of the foreach accessors "
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/worker_set.py:983]: Was thinking the same thing. This message needs more clarification and details, otherwise, we should not hard-deprecate this yet. I'll fix.

File: rllib/env/tests/test_single_agent_gym_env_runner.py (hunk 4 above)
Comment (sven1977) [reply rllib/env/tests/test_single_agent_gym_env_runner.py:25]: Only vie the `config.batch_mode="complete_episodes"` but then you cannot specify the exact number of episodes to sample, correct? UNLESS you are in the eval worker world, where you have the additional settings: `evaluation_duration` (<- num episodes) and `evaluation_duration_unit="episode"`

File: rllib/env/testing/single_agent_gym_env_runner.py (hunk 3 above)
Comment (sven1977) [reply rllib/env/testing/single_agent_gym_env_runner.py:111]: This makes things much simpler now as e.g. in our current RLlib VectorEnv, we have to check individual environments for being terminated and then reset them individually.

File: rllib/env/env_runner.py (hunk 7 above)
Comment (sven1977) [reply rllib/env/env_runner.py:96]: True, I'm not sure either about the integration of RW into this. Maybe we should retro-fix its sample method into this API (it wouldn't break anything). Alternatively, we could just give the RW version `**kwargs` and then warn if a user sets any of these kwargs.

I do feel that these new arg options of the base `EnvRunner` class are important.
* They allow flexibility by supporting overrides to the default behavior imposed by `self.config`.
* Allows for quick testing/evaluation runs of n timesteps (or n episodes), with exploration or w/o.
* In short, it allows me to use my once configured EnvRunner for different scenarios w/o needing a new EnvRunner with a different config. Like right now, I feel like we cannot really use the main training RolloutWorkers for a quick eval check, you need to setup an extra eval worker set for this, which is tedious. Maybe I just want to pause my training iterations and perform a quick n episode test run (no exploration) to see, where I'm at right now with learning progress.

File: rllib/evaluation/rollout_worker.py (hunk 6 above)
Comment (sven1977) [reply rllib/evaluation/rollout_worker.py:878]: fixed.

File: rllib/evaluation/rollout_worker.py (hunk 5 above)
Comment (sven1977) [reply rllib/evaluation/rollout_worker.py:165]: Moved `get_node_ip`, `get_host`, `find_free_port` back into RW-only. Not part of this base EnvRunner API anymore.

File: rllib/env/env_runner.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+                to sample via `self.config`.
+            explore: Whether to use some exploration strategy to determine the actions
+                to send to the env. Only has an effect if `random_actions` is False.
+            random_actions: Whether to act completely randomly in the env. If True, the
+                value of `explore` has no effect.
+            force_reset: If True, will force-reset the env at the very beginning and
+                thus begin sampling from freshly started episodes.
+
+        Returns:
+            A batch of data of any form.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/env_runner.py]: Enhanced the docstring to make this more clear.

File: rllib/env/env_runner.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/env_runner.py:121]: done

File: rllib/env/env_runner.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        max_num_worker_restarts: int = 0,
+    ) -> type:
+        """Returns the EnvRunner class as a `@ray.remote` using given options.
+
+        The returned class can then be used to instantiate ray actors.
+
+        Examples:
+            >>> remote_cls = EnvRunner.as_remote(num_cpus=1, num_gpus=0)
+            >>> array = [remote_cls.remote({"key": "value"}) for _ in range(10)]
+            >>> results = [env_runner.sample.remote() for env_runner in array]
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/env_runner.py]: Fixed (hopefully). Can you check, whether this is ok?

File: rllib/evaluation/worker_set.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/worker_set.py]: done

File: rllib/env/env_runner.py (hunk 1 above)
Comment (sven1977) [reply rllib/env/env_runner.py:11]: updated the description of this PR.

File: rllib/env/env_runner.py (hunk 7 above)
Comment (sven1977) [reply rllib/env/env_runner.py]: This has been done. I made the API even more minimal now. Just `sample(self, **kwargs)`. The rest can be implemented as needed by subclasses.

**Review by kouroshHakha — APPROVED** [review]

**Comment by kouroshHakha** [comment]
@sven1977 The proposed changes LGTM. Approved conditioned on the tests passing.

File: rllib/env/env_runner.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    from the environment(s).
+    - EnvRunner offers parallelism via creating n remote Ray Actors based on this class.
+    Use the `as_remote()` method to create the corresponding Ray remote class. Then
+    instantiate n Actors using the Ray `[ctor].remote(...)` syntax.
+    - EnvRunner clients can get information about the server/node on which the
+    individual Actors are running.
+    """
+
+    @classmethod
+    def as_remote(
```
[the remark this answers is not in the document]
Comment (Rohan138) [reply rllib/env/env_runner.py:29]: I see we're inheriting it from `RolloutWorker`, just curious if we want to deprecate it eventually

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/env_runner.py:29]: Huh, interesting. That would be another great simplification. Let me try this. ...
Thanks @Rohan138 !
